keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,146 +159,157 @@ consistent with the syntax supported by any of the common proxy vendors:
 
 Among cloud providers, there is also inconsistent levels of support
 for regular expression-based path matching. See the load-balancer
-documentation for [aws][aws-re], [gcp][google-re], [azure][azure-re].
+documentation for [AWS][aws-re], [GCP][google-re], [Azure][azure-re].
 
 It is also the case that our [documentation][ingress-docs] (and most
 Ingress providers) treats the path match as a prefix match. For
 example, a narrow interpretation of the specification would require
 all paths to end with `".*$"`.
 
-#### Proposal
-
-1. Clarify behavior of the existing path regex match and identify that use of
-   path regex is non-portable.
-1. Support prefix matching behavior as the portable alternative.
+A detailed discussion of this issue can be found
+[here](https://github.com/kubernetes/ingress-nginx/issues/555).
 
-Change specification of `ingress.spec.rules.http.paths.path` from the current
-text to indicate the non-portability of the value:
+#### Paths proposal
 
-> The matching behavior of Path is implementation specific. Path is either a
-> prefix or regular expression match.
+1. Explicitly state the match mode of the path.
+1. Support the existing implementation-specific behavior.
+1. Support a portable prefix match and future expansion of behavior.
 
-Add a field `ingress.spec.rules.http.paths.pathPrefix`:
+Add a field `ingress.spec.rules.http.paths.pathType` to indicate
+the desired interpretation of the meaning of the `path`:
 
 ```golang
-type HTTPIngressPath struct {
-  ...
-  // PathPrefix matches the HTTP request if the request path begins with
-  // this string. For example, PathPrefix = "/foo" will match "/foo", "/foox",
-  // "/foo/bar".
+ type HTTPIngressPath struct {
+   ...
+  // Path to match against. The interpretation of Path depends on
+  // the value of PathType.
   //
-  // Both Path and PathPrefix cannot be set (non-empty) at the same time.
-  PathPrefix string
-}
-```
-
-Interoperability between v1beta and v1 will work as follows:
+  // Defaults to "/" if empty.
+  //
+  // +Optional
+  Path string
 
-v1beta1 gets the `PathPrefix" field. v1 gets an annotation
-`ingress.kubernetes.io/non-portable-path-semantics`. If we create a v1 object
-from a v1beta1 where `path` was used, we set the annotation. We keep both Path
-and PathPrefix in the API.
+  // PathType determines the interpretation of the Path
+  // matching. PathType can be one of the following values:
+  //
+  // Prefix - matches based on a URL path prefix split
+  //   by '/'. This is not a substring match. For example:
+  //   "/foo/bar" will match "/foo/bar", "/foo/bar/abc", but
+  //   will not match "/foo/barzzz". A Prefix Path of "/" will
```

Comment:
Is this the most common semantic?  It's different from other things I have seen, e.g. Istio and Envoy (I think) would match /foo/barzzz and require 2 rules to express /foo/bar (exact) and /foo/bar/* (prefix).  I though nginx did too but I can't tell from docs...

@aledbf

Reply:
Envoy has a few options to determine the route that it is matching on.

- [RouteMatch prefix](https://github.com/envoyproxy/envoy/blob/b22d2b5cf09f779962cfedaaab24969f384cbc48/api/envoy/api/v2/route/route.proto#L334) validates that the prefix specified matches the beginning of the :path header

- [RouteMatch path](https://github.com/envoyproxy/envoy/blob/b22d2b5cf09f779962cfedaaab24969f384cbc48/api/envoy/api/v2/route/route.proto#L338) is similar to the Exact case that Tim mentioned below. It validates that the path matches the :path header exactly after the query string is removed.

- [RouteMatch regex](https://github.com/envoyproxy/envoy/blob/b22d2b5cf09f779962cfedaaab24969f384cbc48/api/envoy/api/v2/route/route.proto#L352) utilizes the [C++ std::regex](https://en.cppreference.com/w/cpp/regex/ecmascript) semantics to match the entire path specified in the :path header (minus the query bits). This does not work if only a subset of the :path header matches.

- There is also a new safe_regex that will utilize the [Google RE2 library](https://github.com/google/re2/wiki/Syntax). This new option is under development and is a response to some unsafe (normally user provided) regular expressions that [could cause Envoy to crash](https://github.com/envoyproxy/envoy/issues/7728).

Nginx locations also have similar options but rely on [several key symbols for exact vs more fuzzy matching](http://nginx.org/en/docs/http/ngx_http_core_module.html#location)

I'm not sure if this helps much but figured it would at least add some more context in comparison to a few open source implementations.

Reply:
> /foo/bar (exact) and /foo/bar/* (prefix).

Same for nginx http://nginx.org/en/docs/http/ngx_http_core_module.html#location

location = /foo/bar {} -> exact match
location /foo/bar {}    -> /foo/bar/*

Just in case, ingress-nginx does not provide "exact match" locations.

Reply:
>  location /foo/bar {} -> /foo/bar/*

I'm not sure if this is the case.
Nginx matches `/foo/barzzz` in this case. It is a string prefix match and not a path prefix.

Reply:
Add clarification of the behavior

Reply:
GCE will implement as "/foo/bar/*".
Nginx will implement as location "/foo/bar/" and "=/foo/bar". (prefix + equality)
Envoy will use path match.

It does not seem like a good idea to define this as string-based matching "/foobar", as that has fewer use cases and somewhat unexpected behavior.